### ISSUE TITLE
SystemDetails() - 'total bodies' and 'scanned bodies' are zero on game load.

### ISSUE THREAD

**Issue by Darkcyde13** [EDCD/EDDI#1480]
## EDDI version in which issue is found
v3.4.3

## VoiceAttack version in which issue is found (as applicable)
v1.7.8

## Steps to reproduce
1. Go to a star system not currently known by EDSM (not sure if EDSM is relevant though).
2. Perform the 'honk', and scan some bodies if you like.
3. Quit game, EDDI and VA.
4. Load VA, then load your game.

## Expected
SystemDetails() to report the correct number of total bodies in the system, and the correct number of bodies scanned.

## Observed
Frequently (but not every time) the total number of bodies, and scanned bodies, will be zero or less than the real number.
This can also happen simply when jumping into a new undiscovered system (this just happened in my game).

## Investigation
I noticed this a while ago as I'm on my (slow) way back from DW2.  I'm currently in an area of space where nothing has been discovered yet, and doesn't exist in EDSM.  Most of the time, at the end of a gaming session, I'll fully scan a system and line up for my next jump, ready for when I next play, but occasionally I will leave game with a system only half done.

I have the 'Location' event read out the details of the current system when the game loads, to remind me if I had anything left to scan or map before I make the next route jump.  I use SystemDetails() to tell me the total number of bodies using len(), then I go through the normal loop of checking each body in SystemDetails().bodies to see if it's been scanned and mapped or not.  Often, the .bodies seems to be empty, even if a few bodies were found with the honk.

I also have this happen when I've jumped to a new system, and done the 'honk'.  Even without reloading my game, if the honk found several bodies, the SystemDetails().bodies is empty/zero, or has just the star,  until I start scanning.

The only time I can guarantee accurate results, is when the whole system has been fully scanned.  It's like all the scan data is only then finally fully saved for later use, after the 'System scan complete' event.

I think this may be a bug, however, I've also considered it may be a limitation of SystemDetails() when it retrieves the data from wherever it obtains it from, that the source may be wrong, at least until the system has been fully scanned.

As an example, I've just had this on game load:-
"You are near the primary star, in the Hatchio DW-A b57-0 system. This system has not been scanned, with 2 bodies remaining, so no bodies worth mapping have been detected yet."
This system actually has 1 star and 5 other bodies.  I did the honk which scanned the star, and then scanned one body, then quit game, EDDI and VA, then reloaded all.
- "This system has not been scanned," This is only reported if no bodies have been scanned at all, yet I have scanned the star and one other planet.
- "with 2 bodies remaining," This is total bodies minus scanned bodies, showing only 2 bodies counted in the system, with none of them scanned. Yet there are 6 bodies in this system, and two have been scanned.
- "so no bodies worth mapping have been detected yet" This is based on the number of scanned bodies, which it thinks is zero, when it's actually two.

So, SystemDetails() doesn't appear to be working quite as intended.  At the moment, to work around this, I've had to retain some of my old loading and saving of exploration data.  If there is any further testing or data you would like, just let me know.

**Comment by jwvanderbeck** [comment]
I think this might be related to the problems I was having trying to determine if a system already had data in EDSM or not?

In my case I was checking if the number of bodies in the event object during a Discovery Scan event were different from the number of bodies in the system data retrieved from SystemDetails() with the though being that if they differe then it probably means EDSM either doesn't have the system or only has incomplete data.

In reality I found this to behave inconsistently with sometimes flagging systems that EDSM already h ad and sometimes failing to flag systems that EDSM did not already have.

**Comment by Tkael** [comment]
EDDI should be caching bodies and saving them to the sqlite database as you scan them - it doesn't wait until the system scan is complete to start saving data. You could use a database reading app to verify whether the body is being saved correctly after a scan. 

To confirm... this is only occurring for the 1st system you start in after loading the game, correct?

Do you get the correct count with {system.bodies} (skipping `SystemDetails()` and referencing the current star system directly)?

When EDDI first starts running, CPU activity will be elevated while it consumes journal entries from the prior game session. 
- Is this occurring before or after the elevated CPU activity subsides? 
- If before, does waiting until CPU activity subsides make a difference?

**Comment by Darkcyde13** [comment]
I'll run some tests now, to be as sure as I can with what you have asked.  As far as I can remember, it is only the first system, as I am getting EDDI to report on the current system at game load.  I don't usually ask for system info while playing, so I'm not so sure about that aspect.

I've only been using SystemDetails() for this, so it may be that using system.bodies will work better, but I'll see what I can come up with.

**Comment by Darkcyde13** [comment]
Some initial findings so far, both SystemDetails().bodies and system.bodies are giving exactly the same data.

I jumped to a system, did the honk, then quit game and EDDI.  Reloaded EDDI, then started the ED launcher and loaded the game.

The 'Location' event is where I'm running the system reporting, and I've added listing the variables at the end of the script.  This is what I get:-
"Good evening commander, welcome back. You currently have no missions.
You are near the primary star, in the Hatchio VQ-I c24-1 system. This system has been fully scanned, with 1 bodies remaining, but no bodies worth mapping have been detected. Total Bodies: 1. Scanned Bodies: 0. Mapped Bodies: 0. To Map: 0."

This is the same for both SystemDetails().bodies and system.bodies.  However, this star system actually has a total of 7 bodies (1 star and 6 others), with the star being scanned by the honk.  So, it seems that the count of total bodies (by using len(system,bodies) and the number of scanned bodies is incorrect, as these should be 7 and 1 respectively.  Unless the star is deliberately not counted by EDDI as being scanned?

I then scanned two of the 6 planets, and got this on game load (without quitting EDDI):-
"Good evening commander, welcome back. You have no missions at the moment.
You are near the primary star, in the Hatchio VQ-I c24-1 system. This system has been fully scanned, with 1 bodies remaining, but no bodies worth mapping have been detected. Total Bodies: 3. Scanned Bodies: 2. Mapped Bodies: 0. To Map: 0."
As you can see, Total Bodies is only the count of bodies that have been scanned, including stars, and not the total number of bodies in the system, scanned or not, which is what I would have expected.  Scanned  bodies seems to only count planets/moons and not stars.

I've tried reloading only my game a few times, leaving EDDI (both standalone and with VA) loaded so as not to have the high CPU usage in effect, and I get the same thing every time.

I'm going to find a system with multiple stars and see what I get with that.  I'll also look for a system with planets that get discovered by the honk and see what results I get with that too.

**Comment by Darkcyde13** [comment]
Next phase results.

The next system I jumped to had two stars and 3 bodies.  I did the honk to find the two stars and then scanned one planet.  Logged out of game (but not EDDI) and reloaded:
"Good evening commander, welcome back. You have no missions at the moment.
You are near Star A, in the Hatchio RK-K c23-1 system. This system has been fully scanned, but no bodies worth mapping have been detected. Total Bodies: 3. Scanned Bodies: 3. Mapped Bodies: 0. To Map: 0."

This seems to have now counted both stars as being scanned, unlike the previous system.  However, total bodies is still only those scanned, not the system total of 5.

I then quit everything again.  Loaded EDDI and waited for CPU.  Loaded ED and my game:
"Good evening commander, welcome back. You have no missions at the moment.
You are near Star A, in the Hatchio RK-K c23-1 system. This system has not been scanned, with 3 bodies remaining, so no bodies worth mapping have been detected yet. Total Bodies: 3. Scanned Bodies: 0. Mapped Bodies: 0. To Map: 0."

Now, this seems to have totally forgotten that I scanned anything, so scanned bodies wasn't saved, although total bodies is remembered (albeit the wrong number).

I then changed the script to use system.bodies (had been SystemDetails() ) and quit and reloaded everything again, but I got exactly the same result: 3 total, 0 scanned.

Final testing will be a system where bodies are scanned with the honk, but that may take more time to find....

**Comment by jwvanderbeck** [comment]
It is my understanding that SystemDetails() is fetching the data from EDSM not the game.  I looked up system Hatchio RK-K c23-1 and it says it was first discovered (on EDSM) by you, so its possible that it said 3 bodies because at that time that was all that you had found, and thus all that EDSM knew about.

Looking at the system now it shows 5 bodies, but I am assuming that is because you have sense scanned all 5 and moved on.

Might be worth looking at how many bodies EDSM shows when you do a partial scan and then log off.  Assuming its a new system that is.

**Comment by Darkcyde13** [comment]
That makes sense.  I've just checked out my current system.  It has 3 stars and 10 other bodies.  I've only done the honk and EDSM is saying only the three stars, which would be correct.

However, with that system you mention, Hatchio RK-K c23-1, EDDI 'forgot' I'd scanned 3 after quitting and reloading, and it never said I'd scanned all five bodies.

I've also looked at the EDDI database and for that system, the two stars and one planet say "null" for 'scanned', whereas the other two have a timestamp.  There seem to be a lot of other 'null' references, but I'm not sure that they should all have values anyway.  So it would seem that the local DB EDDI is keeping, isn't being written to properly, and thus it's not reporting correctly on game load.  At least that's how it looks to me, but I'm a DB noob, I've not touched sqlite since I was in college 20+ years ago. :/

**Comment by jwvanderbeck** [comment]
> However, with that system you mention, Hatchio RK-K c23-1, EDDI 'forgot' I'd scanned 3 after quitting and reloading, and it never said I'd scanned all five bodies.

Yeah I wasn't trying to explain away all the issues as clearly something isn't working correctly.  I was simply pointing out a possible explanation for why the total bodies seemed off.

**Comment by Darkcyde13** [comment]
I just told EDDI to 'obtain logs' but that didn't correct the null scanned bodies.  I'm not sure if it should or not though.

**Comment by Tkael** [comment]
The `SystemDetails()` function uses cached data from your local database first. If the data is "stale" (typically more than an hour hold) then it'll pull fresh data from EDSM. 

`obtain logs` obtains a list of star system names and visit timestamps, which we merge with data in your local database (or retrieved from EDSM if local data is stale). 

{len(system.bodies)} will only include the bodies that you have scanned, it does not count the total number of bodies in the star system. The `Discovery scan` event includes a total number of bodies in the star system. Consequently, you can check the number of bodies that have been scanned by comparing `{event.bodies}` to `{len(system.bodies)}` in the Discovery scan script.

`reportbody.scanned` and `reportbody.mapped` (which are the values that I'm presuming that you're referencing in your script) are timestamp values. If you're seeing null values for those in your database then I'd guess that the most likely explanation is that the old data with your `scanned` and `mapped` timestamps is being pruned as stale and is being replaced with fresher data from EDSM which doesn't include those same timestamps.

**Comment by Darkcyde13** [comment]
Thank you Tkael, that does seem to mostly fit with what I've been seeing in the majority of cases.

It seems I have been misinterpreting the description in the help: "`bodies` the starsystem's bodies (array of Body objects)".  I thought this meant a complete list of all bodies, even if it was just an array of empty objects, not just only those that have been scanned.  I had thought that having a '.scanned' variable was the only way to tell if a body had been scanned.  I had based my assumptions around these two things, and this whole investigation and your explanation have mostly cleared it all up.

However, I said in the majority of cases.  With the Hatchio RK-K c23-1 system (a fresh undiscovered system), it was most certainly a lot less than an hour from scanning to checking the database and finding the null values for the .scanned timestamp.  I had scanned the stars and one planet, checked the report (which was correct), then quit and reloaded.  It was on game load that EDDI said I hadn't scanned anything (so maybe about 5 minutes after the scan), and not long after that when I checked the DB.  No more than 30-40 minutes in total.  I can only assume that the .scanned data was correct in memory, but for whatever reason, had not been written to the DB.  Yet, the rest of the information for those stars and the planet HAD been written, just not the scanned timestamp.  I'm not sure how this would have been considered as stale data.

I frequently get problems like this on game load, but now I know what to do to correct my scripts and get them working the way I had intended.

I haven't had time to play this evening, but I'll fix my scripts, and try some more testing tomorrow.  I have a feeling I will still come across instances of the DB not having the correct .scanned timestamps, as the symptoms I experience most seem to point to that being the most likely cause of my problems.

So, for the moment, after all this, could I possibly request that when a discovery scan is done, the total number of bodies in the system is also stored in the DB?  Maybe as a .totalbodies decimal?  I would like something that I can reference when I load a game, and something that I can check against when scanning system bodies.  I know that can easily be done with a state variable when the honk is done, but I won't be doing that at game load.  Although, I could just add it into the honk event script I guess, I just think it would be tidier with a variable, plus in discovered systems, people won't really be doing a disc. scan to get that information.

Thanks again. :)

**Comment by Tkael** [comment]
It's all based on the timestamp for when you entered the system. If you scanned something less than an hour ago but entered the system more than an hour ago then it could still be considered stale.

**Comment by Tkael** [comment]
I've created a separate issue for the totalbodies request.

**Comment by Darkcyde13** [comment]
Thanks for the separate issue. :)

As for the null data in the DB, I've gone back through my ED Journals to find the timestamps of the things I did.  I wasn't too far off with my estimations, so I've run another test, making note of all times.

First, here's the timeline for the system mentioned above, Hatchio RK-K c23-1.

> Hatchio RK-K c23-1
> Arrival - 19:31:55
> Two Stars Scanned - 19:32:00 (Auto scan)
> Body A2 Scanned - 19:32:49
> 
> Reloaded Game from Main Menu - 19:33:27
> Shutdown Game - 19:36:53
> 
> Reloaded Game - 19:44:09
> Shutdown Game - 19:47:10
> 
> Reloaded Game - 19:50:08
> Body AB1 Scanned - 20:10:37
> Body A1 Scanned - 20:11:41
> 'All Bodies Found'
> Shutdown Game - 20:12:54

It was soon after this that I checked the EDDI DB and found the null entries for the 'scanned' variables.  As described above I had been quitting and loading ED and EDDI on and off between those times, so I can't say for certain when the DB gained the null entries.

So, to try to work it out, I went to another fresh undiscovered system, and ran through things again.

> Hatchio CM-F b27-0
> Arrival - 20:38:17
> Star Scanned - 20:38:22 (Auto scan)
> Checked EDDI DB - 'scanned' had timestamp
> 
> Reloaded from Main Menu - 20:39:48
> Checked EDDI DB - 'scanned' had timestamp
> 
> Quit Game - 20:42:49
> Checked EDDI DB - 'scanned' had timestamp
> 
> Quit EDDI
> Checked EDDI DB - 'scanned' had timestamp
> 
> Reloaded EDDI
> Checked EDDI DB - 'scanned' was now NULL
> 
> Time from arrival in system < 6 minutes.

So, as you can see, the DB was updated with a null entry for the scanned timestamp when I reloaded EDDI, and all within 5 minutes of entering the system.  This would match the way I've done things before when I experienced these problems: by logging out of ED and EDDI before finishing a full scan of the system, even if I logged back in within a few minutes.

I'm now going to try again with the next system, and see if doing the full scan of all bodies, but not jump out, does the same thing.  I'm curious to see if scanned gets nulled after a complete scan, but before jumping out.

**Comment by Darkcyde13** [comment]
Sigh, back to the drawing board.  I've tried with another system, but can't make it fail like the one above.  I've tried quitting and reloading both the game and EDDI multiple times, but the scanned timestamps always remain correct.  The system does have 33 bodies, so I'm not sure if that could have an effect.

Looks like I'll just have to keep an eye out for it happening again, and see if I can't figure out why. :(

**Comment by jwvanderbeck** [comment]
I doubt reloading the game as any effect on things.  It sounds to me like in **certain** (currently unverified) conditions, EDDI is corrupting the data in its internal database either upon shutdown or startup.

> Quit EDDI
> Checked EDDI DB - 'scanned' had timestamp

Was this the order of operations? So you quit EDDIT then after EDDI was closed, checked the database and the timestamp was valid?

If so then it would seem to me the problem is under some unknown condition at startup the value is getting corrupted.

**Comment by Darkcyde13** [comment]
Yes, that's correct, that was the order I performed each item in.
>If so then it would seem to me the problem is under some unknown condition at startup the value is getting corrupted.

That was my thought too, as it only happened after restarting EDDI, but in the next system it all worked fine.

**Comment by Darkcyde13** [comment]
>Yes, after mapping the DB should fill with valid data. Were there any exceptions recorded in %appdata%/EDDI.log for that session?

I've checked all my EDDI logs, but nothing other than [INFO] entries, and warnings that the VA responder couldn't start where I have been running EDDI as standalone.

>Since EDDI does not retrieve data from EDSM until it is stale (more than an hour since entering the system), 

If this is the case, then how did it happen within a few minutes of entering the system for Hatchio CM-F b27-0, as mentioned above?

**Comment by Tkael** [comment]
I'm not sure. I've got my exploration ship out and I've been trying to test this but so far I haven't been seeing the same results that you have. Both the endpoints for reading and writing to the database are returning a valid `scanned` timestamp (for a system I entered for the first time less than an hour ago).

**Comment by Darkcyde13** [comment]
Whatever it is that's going on, at least some of it seems to be intermittent.  As I mentioned above (4 days ago from this post) I tried to make it go wrong when I arrived in the 33 body system, but just as with your testing, I couldn't make it go wrong then.  When I've experienced the symptoms of this before, it's not been all the time.  Sorry, I know that intermittent faults make diagnosing the problem infinitely more difficult. :(
